src/System.Text.Json/src/System/Text/Json/JsonUtf8Reader.cs

```diff
@@ -0,0 +1,1274 @@
+            return true;
+        }
+
+        private bool ReadFirstToken(byte first)
+        {
+            if (first == JsonConstants.OpenBrace)
+            {
+                _currentDepth++;
+                _stackFreeContainer = 1;
+                _tokenType = JsonTokenType.StartObject;
+                _consumed++;
+                _lineBytePosition++;
+                _inObject = true;
+                _isNotPrimitive = true;
+            }
+            else if (first == JsonConstants.OpenBracket)
+            {
+                _currentDepth++;
+                _tokenType = JsonTokenType.StartArray;
+                _consumed++;
+                _lineBytePosition++;
+                _isNotPrimitive = true;
+            }
+            else
+            {
+                //Create local copy to avoid bounds checks.
+                ReadOnlySpan<byte> localCopy = _buffer;
+
+                if ((uint)(first - '0') <= '9' - '0' || first == '-')
+                {
+                    if (!TryGetNumber(localCopy.Slice(_consumed), out int numberOfBytes))
+                        return false;
+                    _tokenType = JsonTokenType.Number;
+                    _consumed += numberOfBytes;
+                    _lineBytePosition += numberOfBytes;
+                    goto Done;
+                }
+                else if (ConsumeValue(first))
+                    goto Done;
+
+                return false;
+
+            Done:
+                if (_consumed >= (uint)localCopy.Length)
+                    return true;
+
+                if (localCopy[_consumed] <= JsonConstants.Space)
+                {
+                    SkipWhiteSpace();
+                    if (_consumed >= (uint)localCopy.Length)
+                        return true;
+                }
+
+                if (_readerOptions.CommentHandling != JsonCommentHandling.Default)
+                {
+                    if (_readerOptions.CommentHandling == JsonCommentHandling.AllowComments)
+                    {
+                        if (_tokenType == JsonTokenType.Comment || localCopy[_consumed] == JsonConstants.Solidus)
+                            return true;
+                    }
+                    else
+                    {
+                        // JsonCommentHandling.SkipComments
+                        if (localCopy[_consumed] == JsonConstants.Solidus)
+                            return true;
+                    }
+                }
+                ThrowHelper.ThrowJsonReaderException(ref this, ExceptionResource.ExpectedEndAfterSingleJson, localCopy[_consumed]);
+            }
+            return true;
+        }
+
+        private void SkipWhiteSpace()
+        {
+            //Create local copy to avoid bounds checks.
+            ReadOnlySpan<byte> localCopy = _buffer;
+            for (; _consumed < localCopy.Length; _consumed++)
+            {
+                byte val = localCopy[_consumed];
+                if (val != JsonConstants.Space &&
```

Comment:
It would be good to say something like

```C#
// ECMA-404 (2nd edition) section [fill this in] says only these 4 characters count, not all of the Unicode defintions of whitespace.
```

https://github.com/dotnet/corefx/blob/3f763487f688dd826ed56905899082d8d1d68382/src/Common/src/System/Security/Cryptography/AsnReader.cs#L651-L711 is an example of me doing quite a lot of spec-justification on my `if`s.